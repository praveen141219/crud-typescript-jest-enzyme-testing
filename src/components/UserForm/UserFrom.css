.form {
  border: 1px solid;
  padding: 25px;
  display: flex;
  flex-direction: column;
}
.input-parent {
  margin: 10px;
  border: 1px solid;
  padding: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.input-parent label {
  margin-right: 10px;
}
.input-parent input {
  padding: 10px;
  font-size: 15px;
}
button {
  padding: 10px;
  cursor: pointer;
  font-size: 20px;
}
