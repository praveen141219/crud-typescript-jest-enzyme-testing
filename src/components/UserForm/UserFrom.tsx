import { useState } from "react";
import { userFormTypes } from "../../models/userform";
import "./UserFrom.css";
interface formProps {
  usersData: userFormTypes[];
  setUsersData: React.Dispatch<React.SetStateAction<userFormTypes[]>>;
  editedUser: userFormTypes | undefined;
}

const UserForm: React.FunctionComponent<formProps> = ({
  setUsersData,
  usersData,
  editedUser,
}) => {
  const [userFormData, setUserFormData] = useState<userFormTypes>({
    id: 0,
    username: "",
    age: 0,
  });

  const handleChange = (e: React.SyntheticEvent) => {
    editedUser
      ? setUserFormData({
          ...userFormData,
          [(e.target as HTMLInputElement).name]: (e.target as HTMLInputElement)
            .value,
        })
      : setUserFormData({
          ...userFormData,
          id: Math.floor(Math.random() * 100 + 1),
          [(e.target as HTMLInputElement).name]: (e.target as HTMLInputElement)
            .value,
        });
  };

  const formSubmit = () => {
    setUsersData([...usersData, userFormData]);
  };
  const formUpdate = () => {
    console.log(userFormData);
    setUsersData(
      usersData.map((user) =>
        editedUser?.id === user.id
          ? { ...userFormData, user: userFormData }
          : user
      )
    );
  };
  return (
    <>
      <div className="form">
        <div className="input-parent">
          <label>User Name</label>
          <input
            type="text"
            value={editedUser && userFormData.username}
            data-testid="username"
            onChange={handleChange}
            name="username"
          />
        </div>
        <div className="input-parent">
          <label>Age</label>
          <input
            type="number"
            data-testid="age"
            value={editedUser && userFormData.age}
            onChange={handleChange}
            name="age"
          />
        </div>
        <button
          data-testid={editedUser ? "updateBtn" : "submitBtn"}
          onClick={editedUser ? formUpdate : formSubmit}
        >
          {editedUser ? "update" : "submit"}
        </button>
      </div>
    </>
  );
};

export default UserForm;
