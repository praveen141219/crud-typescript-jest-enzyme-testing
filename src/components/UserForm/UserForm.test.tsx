import { shallow } from "enzyme";
import UserForm from "./UserFrom";

describe("user form component", () => {
  const defaultProps = {
    usersData: [],
    setUsersData: () => {},
    setEditedUser: () => {},
    editedUser: undefined,
    isEdit: false,
  };
  const wrapper = shallow(<UserForm {...defaultProps} />);
  test("should  render", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("handle change", () => {
    const onChangeHandle = (testid: string, value: any) => {
      const getChangingInput = wrapper.find(`[data-testid="${testid}"]`);
      return getChangingInput.simulate("change", { target: { value } });
    };
    onChangeHandle("username", "praveen");
    onChangeHandle("age", 25);
  });
  test("handle submit", () => {
    const submitBtn = wrapper.find(`[data-testid="submitBtn"]`);
    submitBtn.simulate("click");
  });
});

describe("isedit true ", () => {
  test("handle update", () => {
    const defaultProps = {
      usersData: [],
      setUsersData: () => {},
      setEditedUser: () => {},
      editedUser: {
        id: 1,
        username: "ddd",
        age: 23,
      },
    };
    const wrapper = shallow(<UserForm {...defaultProps} />);
    const updateBtn = wrapper.find(`[data-testid="updateBtn"]`);
    updateBtn.simulate("click");
    console.log(updateBtn.debug());
  });
});

describe("user form component", () => {
  const defaultProps = {
    usersData: [
      {
        id: 1,
        username: "ddd",
        age: 23,
      },
      {
        id: 2,
        username: "ddd",
        age: 23,
  
      }
    ],
    setUsersData: () => {},
    setEditedUser: () => {},
    editedUser: {
      id: 1,
      username: "ddd",
      age: 23,
    },
    isEdit: false,
  };
  const wrapper = shallow(<UserForm {...defaultProps} />);
  test("should  render", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("handle change", () => {
    const onChangeHandle = (testid: string, value: any) => {
      const getChangingInput = wrapper.find(`[data-testid="${testid}"]`);
      return getChangingInput.simulate("change", { target: { value } });
    };
    onChangeHandle("username", "praveen");
    onChangeHandle("age", 25);
  });
  const updateBtn = wrapper.find(`[data-testid="updateBtn"]`);
  updateBtn.simulate("click");
});
