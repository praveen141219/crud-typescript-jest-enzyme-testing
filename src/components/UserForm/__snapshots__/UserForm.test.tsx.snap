// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user form component should  render 1`] = `
<Fragment>
  <div
    className="form"
  >
    <div
      className="input-parent"
    >
      <label>
        User Name
      </label>
      <input
        data-testid="username"
        name="username"
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="input-parent"
    >
      <label>
        Age
      </label>
      <input
        data-testid="age"
        name="age"
        onChange={[Function]}
        type="number"
      />
    </div>
    <button
      data-testid="submitBtn"
      onClick={[Function]}
    >
      submit
    </button>
  </div>
</Fragment>
`;

exports[`user form component should  render 2`] = `
<Fragment>
  <div
    className="form"
  >
    <div
      className="input-parent"
    >
      <label>
        User Name
      </label>
      <input
        data-testid="username"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="input-parent"
    >
      <label>
        Age
      </label>
      <input
        data-testid="age"
        name="age"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </div>
    <button
      data-testid="updateBtn"
      onClick={[Function]}
    >
      update
    </button>
  </div>
</Fragment>
`;
